nncf/tensorflow/quantization/initializers/collectors.py

```diff
@@ -0,0 +1,276 @@
+"""
+ Copyright (c) 2021 Intel Corporation
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+import numpy as np
+import tensorflow as tf
+
+from nncf.tensorflow.layers.operation import InputType
+from nncf.tensorflow.quantization.initializers.utils import get_per_channel_history
+from nncf.tensorflow.quantization.initializers.utils import discard_zeros
+
+
+class MinMaxStatisticsCollector:
```

Comment:
I wonder if we can reduce the amount of copy-and-paste by using generics to reuse some of the logic between PT and TF stat collectors. Something to the tune of:
```python
class MinMaxStatisticsCollector(Generic[T])
   @abstractmethod
   def min(self) -> T:
          pass   # just as an example

class TFMinMaxStatisticsCollector(MinMaxStatisticsCollector[tf.Tensor])
    def min(self) -> tf.Tensor:
        # implement
```

Reply:
That, if successful, could also allow us to reuse tests. BTW, the unit tests are missing here.

Reply:
Thanks for the comment!

I agree. We have a plan to do so. For now, I provide only functional support and intentionally avoid any generics to better estimate the potential common part for TF and PT stat collectors'.

I am working on the unit tests.

Reply:
@nevget, Please open the ticket on @vshampor's comment. @vshampor Could you resolve conversation?